### ISSUE TITLE
Error thrown in `scheduling_dpmsolver_multistep`  when re-used between validation steps

### ISSUE THREAD

**Issue by EllingtonKirby** [huggingface/diffusers#8711]
I am using `scheduling_dpmsolver_multistep` with 3D point cloud data with a second order update. Because the point clouds contain different numbers of points, I concatenate all the point clouds in a batch into a single cloud. This results in two batches having different numbers of points and thus different sizes.

When stepping the `scheduling_dpmsolver_multistep` during my sampling loop in the validation step, an error is thrown from line [741](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py#L741). The `model_output_list` contains a value from the previous batch in the last position. This previous batch had a different size, which results in an error.

I am a bit confused why the value from the previous batch is being re-used, but I resolved it be re-creating the scheduler object with each new validation step. 

This feels a bit strange, so I wanted to know if this is due to some misuse of the API on my part, or if there is a functionality that could be added to prevent the re-use of `model_outputs` between batches.

**Comment by sayakpaul** [comment]
Could you provide some code snippet that leads to the error? Also, please provide the error trace you're facing. 

Cc: @yiyixuxu

**Comment by EllingtonKirby** [comment]
Sure! 
Here is a very simple mocked version of my code that reproduces the error, and I give a second snippet that avoids the error.

This produces the error:
```
import torch
from diffusers import DPMSolverMultistepScheduler
from tqdm import tqdm

scheduler = DPMSolverMultistepScheduler(
                num_train_timesteps=1000,
                beta_start=3.5e-5,
                beta_end=.007,
                beta_schedule='linear',
                algorithm_type='sde-dpmsolver++',
                solver_order=2,
)
scheduler.set_timesteps(50)

def forward(x_t, t):
    '''
    Mock forward function to illustrate bug
    In the actual code we would be doing a forward pass through the model here.
    '''    
    return torch.randn_like(x_t)


def p_sample_loop(x_t):
    for t in tqdm(range(len(scheduler.timesteps))):
        t = torch.ones(x_t.shape[0]).long() * scheduler.timesteps[t]
        
        noise_t = forward(x_t, t)
        input_noise = x_t

        x_t = scheduler.step(noise_t, t[0], input_noise)['prev_sample']
        
    return x_t

def show_error():
    validation_batch = [torch.randn((100, 3)), torch.randn((101, 3))]
    for example in validation_batch:
        p_sample_loop(example)

show_error()
```
Running this snippet produces the error. 
And here is an example that does not produce the error:
```
def forward(x_t, t):
    '''
    Mock forward function to illustrate bug
    In the actual code we would be doing a forward pass through the model here.
    '''    
    return torch.randn_like(x_t)


def p_sample_loop(x_t):
    scheduler = DPMSolverMultistepScheduler(
                num_train_timesteps=1000,
                beta_start=3.5e-5,
                beta_end=.007,
                beta_schedule='linear',
                algorithm_type='sde-dpmsolver++',
                solver_order=2,
    )
    scheduler.set_timesteps(2)
    for t in tqdm(range(len(scheduler.timesteps))):
        t = scheduler.timesteps[t]
        
        noise_t = forward(x_t, t)
        input_noise = x_t

        x_t = scheduler.step(model_output=noise_t, timestep=t, sample=input_noise)['prev_sample']
        
    return x_t

def no_error():
    validation_batch = [torch.randn((100, 3)), torch.randn((101, 3))] # Note the size difference between the two tensors
    for example in validation_batch:
        p_sample_loop(example)
        
no_error()
```

It looks like the error occurs because in the first `p_sample_loop` call we fall into the `dpm_solver_first_order_update` but the second call we fall into `multistep_dpm_solver_second_order_update`

```
        if self.config.solver_order == 1 or self.lower_order_nums < 1 or lower_order_final:
            prev_sample = self.dpm_solver_first_order_update(model_output, sample=sample, noise=noise)
        elif self.config.solver_order == 2 or self.lower_order_nums < 2 or lower_order_second:
            prev_sample = self.multistep_dpm_solver_second_order_update(self.model_outputs, sample=sample, noise=noise)
        else:
            prev_sample = self.multistep_dpm_solver_third_order_update(self.model_outputs, sample=sample)
```


The project that I based my code on avoids this error by only running for a single batch during validation. I avoid the error by re-initializing the sampler before each sampling loop. But it feels like a hacky way to address the bug. So I was wondering if I am doing something wrong or if there might be a bug here.

Either way thanks for your response! Let me know if I can add any more information.

**Comment by yiyixuxu** [comment]
hi @EllingtonKirby 
you need to call ` scheduler.set_timesteps()` before each denoising loop

**Comment by yiyixuxu** [comment]
marigold pipeline is an example where there are two loops similar to yours, you can see that it calls `set_timesteps` before each inner loop 

https://github.com/huggingface/diffusers/blob/0bae6e447cba0459456c4f7e7e87d7db141d3235/src/diffusers/pipelines/marigold/pipeline_marigold_depth.py#L513

**Comment by EllingtonKirby** [comment]
Okay, thanks for the reference. Will add 👍
